src/ngraph/op/experimental/dyn_pad.cpp

```diff
@@ -0,0 +1,112 @@
+{
+    constructor_validate_and_infer_types();
+}
+
+void op::DynPad::validate_and_infer_types()
+{
+
+    auto arg_t = get_input_element_type(0);
+    auto padding_value_t = get_input_element_type(3);
+    NODE_VALIDATION_CHECK(this,
+                          arg_t.compatible(padding_value_t),
+                          "Padding value and arg type mismatch");
+                          
+    // shape node should have integer data type. For now we only allow i64
+    //TODO: potenially make the type more flexible to include other integer types
+    auto padding_below_et = get_input_element_type(1);
+    NODE_VALIDATION_CHECK(this,
+                          padding_below_et.compatible(element::Type_t::i64),
+                          "DynPad shape must have element type i64, but has ",
+                          padding_below_et);
+
+    auto padding_above_et = get_input_element_type(2);
+    NODE_VALIDATION_CHECK(this,
+                          padding_above_et.compatible(element::Type_t::i64),
+                          "DynPad shape must have element type i64, but has ",
+                          padding_above_et);
+
+    // padding_value is of scalar shape or rank is unknown
+    auto padding_value_rank = get_input_partial_shape(3).rank();
+    NODE_VALIDATION_CHECK(this,
+                          padding_value_rank.is_dynamic() || padding_value_rank.compatible(0),
+                          "DynPad arg is not scalar (rank = 0), but has rank = ",
+                          padding_value_rank);
+    
+    auto arg_shape = get_input_partial_shape(0);
+    auto arg_rank = arg_shape.rank();
+    auto pd_bl_shape = get_input_partial_shape(1);
+    auto pd_bl_rank = pd_bl_shape.rank();
+    auto pd_ab_shape = get_input_partial_shape(2);
+    auto pd_ab_rank = pd_bl_shape.rank();
+    auto output_rank = Rank::dynamic();
+
+    NODE_VALIDATION_CHECK(this, pd_bl_rank.compatible(1), "Partial shape of padding below must be 1");
+    NODE_VALIDATION_CHECK(this, pd_ab_rank.compatible(1), "Partial shape of padding above must be 1");
+
+    if (arg_rank.is_static()) 
+    {
+        // if padding below shape is fully static, check that ranks match
+        NODE_VALIDATION_CHECK(this, 
+                              pd_bl_shape.is_dynamic() || static_cast<Rank>(pd_bl_shape[0]).compatible(arg_rank) , 
+                              "Arg and padding below ranks mismatch");
+
+        NODE_VALIDATION_CHECK(this, 
+                              pd_ab_shape.is_dynamic() || static_cast<Rank>(pd_ab_shape[0]).compatible(arg_rank) , 
+                              "Arg and padding above ranks mismatch");
+        output_rank = arg_rank;                              
+    }
+    else
+    {
+        // arg's rank is dynamic
+        // check if padding above and below have matching ranks
+        
+        NODE_VALIDATION_CHECK(this, 
+                              pd_bl_shape.is_dynamic() || pd_ab_shape.is_dynamic() 
```

Comment:
Keep in mind that `is_dynamic()` does not necessarily mean that the rank is dynamic. (For that you would need to check `pd_bl_shape.rank().is_dynamic()`.) It can mean that rank is static but one or more dimensions are dynamic.